crates/sim2h/src/sim2h_im_state.rs

```diff
@@ -0,0 +1,1177 @@
+    DropConnectionByUri {
+        aol_idx: u64,
+        uri: Lib3hUri,
+    },
+
+    // if this agent is currently assumed connected (&& con_incr is > cur)
+    // mark that they are likely `holding` these aspects
+    AgentHoldsAspects {
+        aol_idx: u64,
+        space_hash: SpaceHash,
+        agent_id: AgentId,
+        entry_hash: EntryHash,
+        aspects: im::HashSet<AspectHash>,
+    },
+
+    // our owner is ready to do a round of gossip
+    // check to see if any of our agents are ready to check
+    // for missing aspects they need to be holding
+    CheckGossip {
+        aol_idx: u64,
+        response: tokio::sync::oneshot::Sender<CheckGossipData>,
+    },
+}
+
+/// we asked for gossip - this indicates which aspects for which agents
+/// should be requested
+#[derive(Debug)]
+pub struct CheckGossipData {
+    pub spaces: im::HashMap<MonoSpaceHash, im::HashSet<MonoAgentId>>,
+}
+
+impl CheckGossipData {
+    pub fn new() -> Self {
+        Self {
+            spaces: im::HashMap::new(),
+        }
+    }
+
+    pub fn add_agent(&mut self, space_hash: MonoSpaceHash, agent_id: MonoAgentId) {
+        let s = self.spaces.entry(space_hash).or_default();
+        s.insert(agent_id);
+    }
+
+    pub fn spaces(self) -> im::HashMap<MonoSpaceHash, im::HashSet<MonoAgentId>> {
+        self.spaces
+    }
+}
+
+#[derive(Clone)]
+struct UpcomingInstant(pub std::time::Instant);
+
+impl std::fmt::Debug for UpcomingInstant {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let now = std::time::Instant::now();
+        let tmp = match self.0.checked_duration_since(now) {
+            Some(d) => format!("{:?} ms", d.as_millis()),
+            None => "[expired]".to_string(),
+        };
+        f.debug_struct("UpcomingInstant").field("in", &tmp).finish()
+    }
+}
+
+impl UpcomingInstant {
+    pub fn new_ms_from_now(ms: u64) -> Self {
+        Self(
+            std::time::Instant::now()
+                .checked_add(std::time::Duration::from_millis(ms))
+                .expect("instant"),
+        )
+    }
+
+    pub fn still_pending(&self) -> bool {
+        let now = std::time::Instant::now();
+        self.0 > now
+    }
+}
+
+/// protocol for sending messages to the `Store`
+#[derive(Debug)]
+enum StoreProto {
+    Mutate(AolEntry, tokio::sync::oneshot::Sender<()>),
+}
+
+/// represents an active connection
+#[derive(Debug, Clone)]
+pub struct ConnectionState {
+    agent_id: MonoAgentId,
+    agent_loc: Location,
+    uri: MonoUri,
+    next_gossip_check: UpcomingInstant,
+}
+
+pub type MonoAgentId = MonoRef<AgentId>;
+pub type MonoSpaceHash = MonoRef<SpaceHash>;
+pub type MonoEntryHash = MonoRef<EntryHash>;
+pub type MonoAspectHash = MonoRef<AspectHash>;
+pub type MonoUri = MonoRef<Lib3hUri>;
+
+/// so we know who's holding what
+pub type Holding = im::HashMap<MonoAspectHash, im::HashSet<MonoAgentId>>;
+
+/// so we cache entry locations as well
+#[derive(Debug, Clone)]
+pub struct Entry {
```

Comment:
calling the `Entry` feels confusing to me.  It seems like it's really `EntryWhereabouts` or `EntryLocationAndHolders` or something like that.  Any better name?

Reply:
Updated to `EntryInfo`